packages/xo-server/src/xapi-stats.js

```diff
@@ -411,40 +211,55 @@ export default class XapiStats {
       .then(response => response.readAll().then(JSON5.parse))
   }
 
-  async _getLastTimestamp (xapi, host, step) {
-    if (this._hosts[host.address][step] === undefined) {
-      const serverTimeStamp = await getServerTimestamp(xapi, host)
-      return serverTimeStamp - step * RRD_POINTS_PER_STEP[step] + step
-    }
+  async _getTimestamp (xapi, host, step) {
+    const serverTimeStamp = await getServerTimestamp(xapi, host.$ref)
+    const maxDuration = step * RRD_POINTS_PER_STEP[step]
+    const lastTimestamp =
+      this._hosts[host.address][step] !== undefined &&
+      this._hosts[host.address][step].endTimestamp
 
-    return this._hosts[host.address][step].endTimestamp
+    if (
+      !lastTimestamp ||
+      serverTimeStamp - lastTimestamp + step > maxDuration
+    ) {
+      return serverTimeStamp - maxDuration + step
+    }
+    return lastTimestamp
   }
 
-  _getPoints (hostname, step, vmId) {
+  _getStats (hostname, step, vmId) {
     const hostStats = this._hosts[hostname][step]
 
-    // Return host points
+    // Return host stats
     if (vmId === undefined) {
       return {
         interval: step,
         ...hostStats,
       }
     }
 
-    const vmsStats = this._vms[hostname][step]
-
-    // Return vm points
+    // Return vm stats
     return {
       interval: step,
       endTimestamp: hostStats.endTimestamp,
-      stats: (vmsStats && vmsStats[vmId]) || getNewVmStats(),
+      stats: this._vms[hostname][step][vmId],
     }
   }
 
-  async _getAndUpdatePoints (xapi, host, vmId, granularity) {
-    // Get granularity to use
+  _parseMemory (collection) {
+    if (collection.memoryFree === undefined) {
+      collection.memoryFree = collection.memory
+    }
+
+    collection.memoryUsed = map(
+      collection.memory,
+      (value, key) => value - collection.memoryFree[key]
+    )
+  }
```

Comment:
Why is there a method specific for a metric?

Reply:
Because it used in `xo-web` and `xo-usage-report`. I tried to have the same result than the old implementation.

Reply:
Because it's a virtual metric, which does not exists by itself in the RRDs?

Reply:
Yes that's true.